bowtie/_cli.py

```diff
@@ -146,53 +146,55 @@ def implementation_subcommand(reporter: _report.Reporter = SILENT):
     """
 
     def wrapper(fn: ImplementationSubcommand):
-        async def run(
-            image_names: list[str],
-            read_timeout_sec: float,
-            make_validator: MakeValidator = make_validator,
-            **kw: Any,
-        ) -> int:
-            exit_code = 0
-            start = _start(
-                image_names=image_names,
-                make_validator=make_validator,
-                reporter=reporter,
-                read_timeout_sec=read_timeout_sec,
+        @subcommand
+        @IMPLEMENTATION
+        @TIMEOUT
+        @wraps(fn)
+        def cmd(image_names: list[str], **kwargs: Any) -> int:
+            return asyncio.run(
+                _run_implementations(
+                    image_names=image_names,
+                    fn=fn,
+                    reporter=reporter,
+                    **kwargs,
+                ),
             )
 
-            running: list[Implementation] = []
-            async with start as implementations:
-                for each in implementations:  # FIXME: respect --quiet
-                    try:
-                        implementation = await each
-                    except StartupFailed as err:
-                        exit_code |= _EX_CONFIG
-                        show: list[console.RenderableType] = [err.diagnostic()]
-                        if err.stderr:
-                            stderr = panel.Panel(err.stderr, title="stderr")
-                            show.append(stderr)
-                        rich.print(*show, file=sys.stderr)
-                        continue
-                    except NoSuchImplementation as err:
-                        exit_code |= _EX_CONFIG
-                        rich.print(err.diagnostic(), file=sys.stderr)
-                        continue
-
-                    running.append(implementation)
-
-                if running:
-                    exit_code |= await fn(implementations=running, **kw) or 0
-                else:
-                    exit_code |= _EX_CONFIG
+        return cmd
+
+    return wrapper
+
 
-            return exit_code
+def all_implementations_subcommand(reporter: _report.Reporter = SILENT):
```

Comment:
(And relatedly then why you introduced this function).

Reply:
My reply on the other comment you gave will help you understand my side of why I introduced this function.